src/site/cpp-objects.rst

```diff
@@ -0,0 +1,170 @@
+**************************
+C++ Lessons #6 --- Objects
+**************************
+
+.. warning::
+
+    * For these, I will only go over the C++ details at a high level
+    * Getting into the nitty gritty of C++ is not the purpose of this aside
+    * Some of the concepts require knowledge of more advanced ideas that have not been covered yet
+        * If this happens, do not panic
+        * it probably makes more sense to revisit this later in the course
+    * For simplicity, some of the examples will use less than ideal implementations
+
+
+* The same example described in :doc:`Topic3 </topic3.rst>` is used here
+    * A collection of ``Friend`` objects stored in an ``ContactList`` obejct
+
+
+Friend
+======
+
+* We will create a ``Friend`` class to keep track of a few details
+    * First Name
+    * Last Name
+    * Email
+
+* We will also want
+    * Constructor
+    * Accessors
+
+* That's about it
+
+
+Header
+------
+
+* The first thing we want is a *header* file
+* Call this ``Friend.h``
+* This is where we typically declare all the member fields and methods the class will have
+* This file will also have all our ``#include``s too
+* Below is a complete ``Friend.h``
+
+
+.. code-block:: cpp
+    :linenos:
+
+    // Friend.h
+    #include <string>
+
+    class Friend {
+        private:
+            std::string firstName_;
+            std::string lastName_;
+            std::string email_;
+
+    public:
+        Friend();
+        Friend(std::string, std::string, std::string);
+
+        const std::string& firstName()  {return firstName_;}
+        const std::string& lastName()   {return lastName_;}
+              std::string& email()      {return email_;}
```

Comment:
This is still not correct. `email` most so.

```
// return a const reference to the property when the class instance itself is const (the 2nd const)
const T& prop() const { return prop_; }

// return a mutable reference when the class is mutable
T& prop() { return prop_; }
```

You need both. Only including the latter means you'll be trying to return a mutable reference to a `const` object -- I'd hope that'd be a compiler error but who knows.

Reply:
For immutable fields, you can omit the latter. C++ will gladly call the `const` overload when an instance is not `const`, but it won't do the other way around.

Reply:
Maybe it's my compiler, but it was pissy about overloaded methods with same return type, regardless of constant.

Reply:
Regardless, I'll investigate further... later

Reply:
[works on my machine](https://github.com/twentylemon/flow/blob/main/flow/flow/optional.h#L129-L151)

Reply:
would've worked on mine too. I was missing the 2nd `const` in each.